text/0225-configuration.md

```diff
@@ -0,0 +1,410 @@
+        },
+      },
+      "propagators": []string{"tracecontext", "baggage", "b3multi"},
+      "tracer_provider": config{
+        "exporters": config{
+          "zipkin": config{
+            "endpoint": "http://localhost:9411/api/v2/spans",
+            "timeout":  10000,
+          },
+          "otlp": config{},
+        },
+        "span_processors": []config{
+          {
+            "name": "batch",
+            "args": config{
+              "schedule_delay":        5000,
+              "export_timeout":        30000,
+              "max_queue_size":        2048,
+              "max_export_batch_size": 512,
+              "exporter":              "zipkin",
+            },
+          },
+        },
+      },
+      ...
+    },
+  },
+)
+```
+
+### Configuration file
+
+The following demonstrates an example of a configuration file format (full example [here](https://github.com/MrAlias/otel-schema/blob/main/config.yaml)):
+
+```yaml
+# include version specification in configuration files to help with parsing and schema evolution.
+scheme_version: 0.1
+sdk:
+  # Disable the SDK for all signals.
+  #
+  # Boolean value. If "true", a no-op SDK implementation will be used for all telemetry
+  # signals. Any other value or absence of the variable will have no effect and the SDK
+  # will remain enabled. This setting has no effect on propagators configured through
+  # the OTEL_PROPAGATORS variable.
+  #
+  # Environment variable: OTEL_SDK_DISABLED
+  disabled: false
+  # Configure resource attributes and resource detection for all signals.
+  resource:
+    # Key-value pairs to be used as resource attributes.
+    #
+    # Environment variable: OTEL_RESOURCE_ATTRIBUTES
+    attributes:
+      # Sets the value of the `service.name` resource attribute
+      #
+      # Environment variable: OTEL_SERVICE_NAME
+      service.name: !!str "unknown_service"
+  # Configure context propagators. Each propagator has a name and args used to configure it. None of the propagators here have configurable options so args is not demonstrated.
+  #
+  # Environment variable: OTEL_PROPAGATORS
+  propagators: [tracecontext, baggage]
+  # Configure the tracer provider.
+  tracer_provider:
+    # Span exporters. Each exporter key refers to the type of the exporter. Values configure the exporter. Exporters must be associated with a span processor.
+    exporters:
+      # Configure the zipkin exporter.
+      zipkin:
+        # Sets the endpoint.
+        #
+        # Environment variable: OTEL_EXPORTER_ZIPKIN_ENDPOINT
+        endpoint: http://localhost:9411/api/v2/spans
+        # Sets the max time to wait for each export.
+        #
+        # Environment variable: OTEL_EXPORTER_ZIPKIN_TIMEOUT
+        timeout: 10000
+    # List of span processors. Each span processor has a name and args used to configure it.
+    span_processors:
+      # Add a batch span processor.
+      #
+      # Environment variable: OTEL_BSP_*, OTEL_TRACES_EXPORTER
+      - name: batch
```

Comment:
What about custom components? I'm suspecting custom schemas will have to be offered from users extending functionality, e.g. `CustomSampler` or `CachedSpanProcessor` mapping to `custom` and `cached` respectively (specially useful for at-runtime-loaded components). If this was already considered, an example would be great.

Reply:
Supporting custom components (processors, exporters, samplers, etc) is [one of the goals](https://github.com/open-telemetry/oteps/pull/225/files#diff-bd4bde8aa8b94a5c3332731193eb54f01b2000c861620e7ec6263d382a4b5979R33).

The syntax for referencing a custom processor would likely just be:
```
sdk:
  tracer_provider:
    span_processors:
      - name: my-custom-processor
         args:
           foo: bar
           baz: qux
```
I think how exactly custom components are loaded with the configuration arguments is going to vary on a language by language basis.

In java, we publish [SPIs](https://github.com/open-telemetry/opentelemetry-java/tree/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi) for providing custom components which can be referenced by name during configuration. For example, users can implement [ConfigurableSpanExporterProvider](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java) and reference the provided name in `OTEL_TRACES_EXPORTER`. 

I imagine we'd use a similar mechanism for file based config. We could pass the args (i.e. `{foo: bar, baz: qux}`) via the `ConfigProperties` passed to method used to [create](https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk-extensions/autoconfigure-spi/src/main/java/io/opentelemetry/sdk/autoconfigure/spi/traces/ConfigurableSpanExporterProvider.java#L23) the component.

Reply:
Added an example of a custom processor as per @jack-berg's example.